src/dmd/interop.d

```diff
@@ -0,0 +1,31 @@
+/+
+Example usage:
+extern (C++) Dstring funLib(Dstring input)
+{
+    import std.string : toUpper;
+    import std.conv : to;
+    return input.toNative.toUpper.to!Dstring;
+}
++/
+
+extern (C++) struct Darray(T)
+{
+    size_t length;
+    T* ptr;
+
+    extern (D) this(T[] a)
+    {
+        this.length = a.length;
+        this.ptr = a.ptr;
+    }
+
+    extern (D) T[] toNative()
+    {
+        // TODO: could even use a raw cast since ABI is same
+        return ptr[0 .. length];
+    }
+}
+
+alias Dstring = Darray!char;
+alias Dcstring = Darray!(const(char));
```

Comment:
`alias string  = immutable(char)[];`, (though imho it should have been const(char), but I guess it's way too late to change that.)

Reply:
(resolved privately on slack https://dlang.slack.com/archives/D8ZMKRGUV/p1521139433000456)
TLDR: immutable has no C++ equivalent

Reply:
The slack conversation is private, so the link isn't helpful to other people.

in short: `pragma(mangle)` doesn't work in `alias`. It has been proposed in https://github.com/dlang/dmd/pull/7458, but closed due to inactivity :/

Reply:
not only that but `pragma(mangle)` wouldn't help with arbitrary array types even if it were implemented. this PR aims at supporting arbitrary D arrays, not just `const(char)[]` for example